test/e2e/delay_restart_e2e_test.go

```diff
@@ -0,0 +1,882 @@
+			t.Fatalf("Failed to update ConfigMap: %v", err)
+		}
+
+		// Verify restart is delayed
+		startTime := time.Now()
+		restartTime := waitForDelayedDeploymentRestart(ctx, t, clients.k8sClient, "", "ConfigMap update")
+		elapsed := time.Since(startTime)
+
+		if elapsed < 4*time.Second { // Allow some margin
+			t.Errorf("Restart happened too quickly: %v (expected ~5s delay)", elapsed)
+		}
+
+		if elapsed > 10*time.Second { // Upper bound to catch issues
+			t.Errorf("Restart took too long: %v (expected ~5s delay)", elapsed)
+		}
+
+		t.Logf("SUCCESS: Delayed restart completed after %v with restart time: %s", elapsed, restartTime)
+	})
+
+	// Test 2: Rule with zero delay
+	t.Run("RuleWithZeroDelay", func(t *testing.T) {
+		t.Log("Testing rule with zero delay (immediate restart)...")
+
+		delay := int32(0)
+		ruleName := "zero-delay-rule"
+
+		rule := createDelayRestartRule(ruleName, delay, "test-config-2", "Deployment", "test-app")
+		if err := clients.k8sClient.Create(ctx, rule); err != nil {
+			t.Fatalf("Failed to create zero-delay RestartRule: %v", err)
+		}
+		defer cleanupRestartRule(ctx, t, clients.k8sClient, ruleName)
+
+		if err := waitForRestartRuleReady(ctx, clients.k8sClient, delayTestNamespace, ruleName); err != nil {
+			t.Fatalf("Zero-delay rule did not become ready: %v", err)
+		}
+
+		// Update ConfigMap to trigger immediate restart
+		configMap := &corev1.ConfigMap{}
+		key := client.ObjectKey{Namespace: delayTestNamespace, Name: "test-config-2"}
+		if err := clients.k8sClient.Get(ctx, key, configMap); err != nil {
+			t.Fatalf("Failed to get ConfigMap: %v", err)
+		}
+
+		configMap.Data["test-key"] = fmt.Sprintf("updated-value-%d", time.Now().Unix())
+		if err := clients.k8sClient.Update(ctx, configMap); err != nil {
+			t.Fatalf("Failed to update ConfigMap: %v", err)
+		}
+
+		// Verify restart is immediate (handled by DelayedRestartManager)
+		startTime := time.Now()
+		restartTime := waitForDelayedDeploymentRestart(ctx, t, clients.k8sClient, "", "ConfigMap update with zero delay")
+		elapsed := time.Since(startTime)
+
+		if elapsed > 3*time.Second { // Should be almost immediate
+			t.Errorf("Zero-delay restart took too long: %v", elapsed)
+		}
+
+		t.Logf("SUCCESS: Zero-delay restart completed after %v with restart time: %s", elapsed, restartTime)
+	})
+}
+
+// testMultipleRulesDelayPriority tests delay priority with multiple rules
+func testMultipleRulesDelayPriority(ctx context.Context, t *testing.T, clients *testClients) {
+	t.Log("=== Testing Multiple Rules Delay Priority ===")
+
+	// Test 1: Multiple rules with different delays - highest should win
+	t.Run("HighestDelayWins", func(t *testing.T) {
+		t.Log("Testing multiple rules with different delays...")
+
+		rule1Name := "delay-rule-3s"
+		rule2Name := "delay-rule-8s"
+		rule3Name := "delay-rule-5s"
+
+		// Create rules with different delays for the same target
+		rule1 := createDelayRestartRule(rule1Name, 3, "test-config-multi-1", "Deployment", "test-app")
```

Comment:
Reusing the deployment previously used in other test cases will not work because it will already contain a restarted At annotation.